pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,255 @@
+	case PatchEnvelopesVsChActionDelete:
+		pes.deleteExternalPatches(volumeStatus.Vs)
+	}
+}
+
+func (pes *PatchEnvelopes) updateExternalPatches(vs types.VolumeStatus) {
+	if vs.State < types.CREATED_VOLUME {
+		return
+	}
+
+	pes.Lock()
+	defer pes.Unlock()
+
+	volumeExists := false
+	for i := range pes.completedVolumes {
+		if pes.completedVolumes[i].VolumeID == vs.VolumeID {
+			pes.completedVolumes[i] = vs
+			volumeExists = true
+			break
+		}
+	}
+	if !volumeExists {
+		pes.completedVolumes = append(pes.completedVolumes, vs)
+	}
+
+	if err := pes.processVolumeStatus(vs); err != nil {
+		pes.log.Errorf("Failed to update external patches %v", err)
+	}
+}
+
+func (pes *PatchEnvelopes) deleteExternalPatches(vs types.VolumeStatus) {
+	pes.Lock()
+	defer pes.Unlock()
+
+	i := 0
+	for _, vol := range pes.completedVolumes {
+		if vol.VolumeID == vs.VolumeID {
+			break
+		}
+		i++
+	}
+
+	pes.completedVolumes[i] = pes.completedVolumes[len(pes.completedVolumes)-1]
+	pes.completedVolumes = pes.completedVolumes[:len(pes.completedVolumes)-1]
+}
+
+func (pes *PatchEnvelopes) updateEnvelopes(peInfo []types.PatchEnvelopeInfo) {
+	pes.Lock()
+	defer pes.Unlock()
+
+	pes.envelopes = peInfo
+	if err := pes.processVolumeStatusList(pes.completedVolumes); err != nil {
+		pes.log.Errorf("Failed to update external patches %v", err)
+	}
+}
+
+func (pes *PatchEnvelopes) processVolumeStatus(vs types.VolumeStatus) error {
+	for i := range pes.envelopes {
+		for j := range pes.envelopes[i].VolumeRefs {
+			volUUID, err := uuid.FromString(pes.envelopes[i].VolumeRefs[j].ImageID)
+			if err != nil {
+				return err
+			}
+			if volUUID == vs.VolumeID {
+				blob, err := blobFromVolRef(pes.envelopes[i].VolumeRefs[j], vs)
+				if err != nil {
+					return err
+				}
+
+				if idx := types.CompletedBinaryBlobIdxByName(pes.envelopes[i].BinaryBlobs, blob.FileName); idx != -1 {
+					pes.envelopes[i].BinaryBlobs[idx] = *blob
+				} else {
+					pes.envelopes[i].BinaryBlobs = append(pes.envelopes[i].BinaryBlobs, *blob)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (pes *PatchEnvelopes) processVolumeStatusList(volumeStatuses []types.VolumeStatus) error {
+	for _, vs := range volumeStatuses {
+		if err := pes.processVolumeStatus(vs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func blobFromVolRef(vr types.BinaryBlobVolumeRef, vs types.VolumeStatus) (*types.BinaryBlobCompleted, error) {
+	sha, err := fileutils.ComputeShaFile(vs.FileLocation)
+	if err != nil {
+		return nil, err
+	}
+	return &types.BinaryBlobCompleted{
+		FileName:         vr.FileName,
+		FileMetadata:     vr.FileMetadata,
+		ArtifactMetadata: vr.ArtifactMetadata,
+		FileSha:          fmt.Sprintf("%x", sha),
+		URL:              vs.FileLocation,
+	}, nil
+}
+
+// PatchEnvelopeInfo contains fields that we don't want to expose to app instance (like AllowedApps), so we use
+// peInfoToDisplay and patchEnvelopesJSONFOrAppInstance to marshal PatchEnvelopeInfoList in a format, which is
```

Comment:
What is `JSONF`?

Reply:
Typo and yetus format. JSON For App Instance. Fixed